src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -24,133 +24,124 @@ trait Batchable {
-        else {
-          val newOther = new Array[Runnable](len << 1)
-          System.arraycopy(other, 0, newOther, 0, len)
-          newOther
-        }
+    private[this] final def ensureCapacity(curSize: Int): Array[Runnable] = {
+      val curOther = this.other
+      val curLen = curOther.length
+      if (curSize <= curLen) curOther
+      else {
+        val newLen = if (curLen == 0) 4 else curLen << 1
+
+        if (newLen <= curLen) throw new StackOverflowError("Space limit of asynchronous stack reached: " + curLen)
+        val newOther = new Array[Runnable](newLen)
+        System.arraycopy(curOther, 0, newOther, 0, curLen)
+        this.other = newOther
+        newOther
+      }
     }
 
     final def push(r: Runnable): Unit = {
-      val sz = size
-      if(sz > 0) {
-        if (sz > other.length)
-          grow()
-        other(sz - 1) = r
-      } else first = r
-      size = sz + 1
+      val sz = this.size
+      if(sz == 0)
+        this.first = r
+      else
+        ensureCapacity(sz)(sz - 1) = r
+      this.size = sz + 1
     }
 
-    final def pop(): Runnable =
-      (size: @switch) match {
-        case 0 => null
-        case 1 =>
-          val ret = first
-          first = null
-          size = 0
-          ret
-        case n =>
-          val ret = other(n - 2)
-          other(n - 2) = null
-          size = n - 1
-          ret
+    final def pop(): Runnable = {
+      val sz = this.size
+      if (sz < 2) {
+        val ret = this.first
+        this.first = null
+        this.size = 0
+        ret
+      } else {
+        val o = this.other
+        val ret = o(sz - 2)
+        o(sz - 2) = null
+        this.size = sz - 1
+        ret
       }
+    }
 
     // this method runs in the delegate ExecutionContext's thread
     override final def run(): Unit = {
-      //This invariant needs to hold: require(_tasksLocal.get eq null)
       _tasksLocal.set(this)
-      val failure = BlockContext.usingBlockContext(this)(this)
+
+      val failure = // Only install the block context if we can resubmit on blocking
+        if (resubmitOnBlock) BlockContext.usingBlockContext(this)(this)
+        else runWithoutResubmit(runUntilFailureOrDone())
+
       _tasksLocal.remove()
-      if (failure ne null)
+      if (failure != null)
         throw handleRunFailure(failure)
     }
 
     override final def apply(prevBlockContext: BlockContext): Throwable = {
       parentBlockContext = prevBlockContext
-      var failure: Throwable = null
-      try {
-        var r = pop()
-        while(r ne null) {
-          r.run()
-          r = pop()
-        }
-      } catch {
-        case t: Throwable => failure = t
-      }
+      val failure = runUntilFailureOrDone()
       parentBlockContext = null
       failure
     }
 
+    @tailrec private[this] final def runWithoutResubmit(failure: Throwable): Throwable =
+      if ((failure ne null) && (failure.isInstanceOf[InterruptedException] || NonFatal(failure))) {
```

Comment:
if we're being consistent with your above change,

```suggestion
      if ((failure != null) && (failure.isInstanceOf[InterruptedException] || NonFatal(failure))) {
```

Reply:
Agreed!